streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -272,15 +280,22 @@ public synchronized void onChange(final StreamThread thread,
             }
             if (newState == StreamThread.State.PARTITIONS_REVOKED ||
                 newState == StreamThread.State.ASSIGNING_PARTITIONS) {
-                setState(State.REBALANCING);
+                setStateWhenNotShuttingDownOrNotRunning(State.REBALANCING);
             } else if (newState == StreamThread.State.RUNNING) {
+                // one thread is running, check others
                 for (final StreamThread.State state : threadState.values()) {
                     if (state != StreamThread.State.RUNNING) {
                         return;
                     }
                 }
-                setState(State.RUNNING);
+                setStateWhenNotShuttingDownOrNotRunning(State.RUNNING);
+            } else if (newState == StreamThread.State.DEAD) {
+                // one thread died, check if we have enough threads running
+                if (threadState.size() == 0) {
+                    setStateWhenNotShuttingDownOrNotRunning(State.NOT_RUNNING);
```

Comment:
Is this correct? Should we set it to `NOT_RUNNING` even if the current status is `PENDING_SHUTDOWN`?